File: leetcode/lc/T401-450/T437.cc
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int pathSum(TreeNode *root, int sum)
    {

        map<int, int> m;
        m[0] = 1;

        return travel(root, m, 0, sum);
    }
    int travel(TreeNode *root, map<int, int> &m, int cur, int sum)
    {
        if (root == nullptr)
            return 0;
        int res = 0;
        cur += root->val;
        res += m[cur - sum];
        m[cur] += 1;
        res += travel(root->left, m, cur, sum);
        res += travel(root->right, m, cur, sum);
        m[cur]--;
        return res;
    }
};

File: leetcode/lc/T401-450/t_437.java
package leetcoder;


import java.util.*;



class Solution {

    public int pathSum(TreeNode root, int sum) {
        int count = 0;
        Stack<TreeNode> s = new Stack<>();
        s.push(root);
        while(!s.isEmpty()){
            TreeNode pop = s.pop();
            count += sum(pop,sum);
            if(pop.left != null)
                s.push(pop.left);
            if(pop.right!=null)
                s.push(pop.right);
        }
        return count;
    }

    private int sum(TreeNode pop, int sum) {
        int count = 0;
        if(pop == null) return 0;
        if(sum == pop.val) count++;
        count+=sum(pop.left,sum-pop.val);
        count+=sum(pop.right,sum-pop.val);
        return count;


    }
}



public class AAA{
    public static void main(String[] args) {

    }
}